lib/remap-functions/src/parse_key_value.rs

```diff
@@ -0,0 +1,370 @@
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+            Parameter {
+                keyword: "separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let field_split = arguments.optional("field_split").map(Expr::boxed);
+        let separator = arguments.optional("separator").map(Expr::boxed);
+
+        Ok(Box::new(ParseKeyValueFn {
+            value,
+            field_split,
+            separator,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ParseKeyValueFn {
+    value: Box<dyn Expression>,
+    field_split: Option<Box<dyn Expression>>,
+    separator: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseKeyValueFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let bytes = self.value.execute(state, object)?.try_bytes()?;
+        let value = String::from_utf8_lossy(&bytes);
+
+        let field_split = match &self.field_split {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => "=".to_string(),
+        };
+
+        let separator = match &self.separator {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => " ".to_string(),
+        };
+
+        let (_, values) =
+            parse_line(&value, &field_split, &separator).map_err(|e| e.to_string())?;
+
+        Ok(values.into_iter().collect::<BTreeMap<_, _>>().into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .merge_optional(
+                self.field_split
+                    .as_ref()
+                    .map(|field_split| field_split.type_def(state)),
+            )
+            .merge_optional(
+                self.separator
+                    .as_ref()
+                    .map(|separator| separator.type_def(state)),
+            )
+            .into_fallible(true)
+            .with_constraint(value::Kind::Map)
+    }
+}
+
+/// Parse the line as a separated list of key value pairs.
+fn parse_line<'a>(
+    input: &'a str,
+    field_split: &'a str,
+    separator: &'a str,
+) -> IResult<&'a str, Vec<(String, Value)>> {
+    separated_list1(
+        parse_separator(separator),
+        parse_key_value(field_split, separator),
+    )(input)
+}
+
+/// Parses the separator between the key/value pairs.
+/// If the separator is a space, we parse as many as we can,
+/// If it is not a space eat any whitespace before our separator as well as the separator.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
```

Comment:
Any idea about this?

If I remove the lifetimes I get the following error, so it seems they are needed: 

```
error[E0759]: `separator` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
```

Reply:
cc @fanatid @pablosichert

Reply:
```rust
fn parse_separator(separator: &str) -> impl Fn(&str) -> IResult<&str, &str> + '_
```

should do the trick.

It's mentioned in https://rust-lang.github.io/rust-clippy/master/#needless_lifetimes under _Known problems_: 

> Lifetime bounds such as impl Foo + 'a and T: 'a must be elided with the placeholder notation '_ because the fully elided notation leaves the type bound to 'static.

As to why, I also don't know exactly. Gotta read up on the `+` operator for lifetimes.

Reply:
@pablosichert Awesome, thanks! I also have some reading to do.